boa/src/exec/mod.rs

```diff
@@ -409,3 +414,43 @@ impl Executable for Node {
         }
     }
 }
+
+/// Store a builtin constructor (such as Object) and its corresponding prototype field
+#[derive(Debug)]
+pub struct StandardConstructor {
+    pub constructor: GcObject,
+    pub prototype: GcObject,
+}
+
+impl Default for StandardConstructor {
+    fn default() -> Self {
+        Self {
+            constructor: GcObject::new(Object::default()),
+            prototype: GcObject::new(Object::default()), // __proto__ === null
+        }
+    }
+}
```

Comment:
There is a problem with this when we initialize Standard Objects we create some dummy garbage collected objects on the need then we immediatly drop them when we assign to them. This is inefficient.

Reply:
True. Is there any way to then initialize the StandardObjects structure. Since we are putting it in the `Interpreter`, we need to initialize it before initializing the builtins.
Using `Option<StandardConstructor>` is an option. But subsequent accesses would require unwrapping the `Option`, which is an unnecessary check.

Reply:
> True. Is there any way to then initialize the StandardObjects structure. Since we are putting it in the `Interpreter`, we need to initialize it before initializing the builtins.
> Using `Option<StandardConstructor>` is an option. But subsequent accesses would require unwrapping the `Option`, which is an unnecessary check.

Hmmm. yep that is a problem. how about we initialize them inplace so the default the `make_constructor_fn` instead of returning `GcObject` we pass the constructor.
The second option is to go with `Option` which gives unnecessary check but it is possible to remove this check with a little unsafe code. with https://doc.rust-lang.org/stable/std/hint/fn.unreachable_unchecked.html

What so you think of this? @54k1 @Razican

Reply:
The question is, why do we need `StandardConstructor` to implement `Default`? Can we do it in a different way?

If there is no other option, and we know this will only be an optional thing on creation as default, what about a builder pattern? We create a `StandardConstructorBuilder`, and use an `Option`. Then we call `build()` which returns a `StandardConstructor` that has no optional field.

That would avoid using `unsafe`. But if we really want to use `unsafe`, using [`MaybeUninit`](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html) would reduce the pointer size a bit, compared to `Option`.

Reply:
@Razican, we need the default since we are embedding the `StandardObjects` in the `Interpreter` which must be initialized. About the builder pattern, how would you access newly created objects?

@HalidOdat, the default allocation is redundant, but there would not be too many allocations such that this becomes a bottleneck. What do you think?

Sorry about the late response. Had some school stuff to do.

Reply:
> @HalidOdat, the default allocation is redundant, but there would not be too many allocations such that this becomes a bottleneck. What do you think?

It's probably not that big of a bottleneck, but I would like for us to find a nicer way of doing this in the future but for now I guess it fine :)